### DISCUSSION TITLE
Flip the screen?

### DISCUSSION THREAD

**Discussion by MarkusNiewoehner** [KlipperScreen/KlipperScreen#120] [General]
Hello!
it is possible to rotate the screen 180 ° ? 

Kind regards,
Markus

**Comment by HermanGaviriaMarin** [comment]
I was looking for this same solution. Turns out that it should be done as per the manufacturers directions, that is, find the official website of documentation of your specific screen and it is very possible that the manufacturer has some SSH commands you can run on putty to rotate or flip the screen. You maybe have to change klipper screen to make sure that touch input registers but thats a maybe.

A quick example of what I mean is here on my displays Github:

https://github.com/goodtft/LCD-show#:~:text=How%20to%20rotate,the%20latest%20drivers)

here they explain two ways flipping it, but remember that these directions will probably not work for you unless you have the same screen as I do, which is the
https://www.amazon.com/Elecrow-800x480-Interface-Supports-Raspberry/dp/B013JECYF2/ref=sr_1_1_sspa?keywords=5+inch+tft&qid=1639335478&sr=8-1-spons&psc=1&spLa=ZW5jcnlwdGVkUXVhbGlmaWVyPUEzSTY2RUlPV1lFMEhNJmVuY3J5cHRlZElkPUEwMjUzMTE2V1hMVDJGWDA5UEFaJmVuY3J5cHRlZEFkSWQ9QTAwNDQ5NDVTWEJLSDgwN1ZZTFUmd2lkZ2V0TmFtZT1zcF9hdGYmYWN0aW9uPWNsaWNrUmVkaXJlY3QmZG9Ob3RMb2dDbGljaz10cnVl,

**Comment by Sirnut** [comment]
Took me longer than I care to admit that all it took was adding lcd_rotate=2 to the [all] section

**Reply by klotzma** [reply]
I also want to rotate the display.  
Where did you put that with the [all]?  
In the klipperscreen config ?

**Reply by peedubwoo** [reply]
That only works if you're using a screen with a raspberrypi.

I've managed to rotate KlipperScreen for an old x86 tablet that was defaulting to portrait mode by editing xorg.conf to rotate the monitor, and changing to coordinate transformation matrix in xinput to rotate touchscreen inputs.

**Comment by downace** [comment]
I'm using Raspberry Pi 4B with BTT Pi TFT50 V2.0, was able to rotate the screen by adding `display_lcd_rotate=2` to `/boot/config.txt`, section `[all]`.

Found this solution [here](https://www.reddit.com/r/BIGTREETECH/comments/w2asxc/rotate_btt_pitft50_v2_for_portrait_mode_display/), it mentions [the official documentation page](https://www.raspberrypi.com/documentation/computers/configuration.html#legacy-graphics-driver).

Touchscreen rotation is also required, [solution from KlipperScreen documentation](https://klipperscreen.readthedocs.io/en/latest/Hardware/#touchscreen-touch-rotation) worked as expected.

**Comment by tahsin3323** [comment]
i had te same Problem with my pi4b and pi 7" lcd

this is how i solved :

>open putty

login

>sudo nano /boot/config.txt

add to one of the lines

>lcd_rotate=2

save and exit(for the save)

>ctrl+x

>Y

>enter

>sudo reboot

and DONE

**Comment by agorafalaht** [comment]
Hello everyone, I have a board from MKS SKIPR with LCD MKS PI TS35, I'm trying to rotate the screen 90° to be able to use it in the best way in my printer, but I haven't found anything that works to rotate the screen. Someone already did. ?Thank you all.

**Reply by tahsin3323** [reply]
> Hello everyone, I have a board from MKS SKIPR with LCD MKS PI TS35, I'm trying to rotate the screen 90° to be able to use it in the best way in my printer, but I haven't found anything that works to rotate the screen. Someone already did. ?Thank you all.

Did you try the codes that above?
Did you connect your pi with 'Putty' and edit your config.txt?

**Reply by agorafalaht** [reply]
yes I already tried, but this config.txt does not exist on the mks pi and mks skipr board

**Reply by Hassla** [reply]
Did you mangage to resolve this issue?

**Reply by Hassla** [reply]
https://github.com/makerbase-mks/MKS-PI/blob/main/README.md#how-to-rotate-180-for-the-ts35-display

So this worked for me!

**Reply by nesickc** [reply]
The link above is for 180°. Is there a working option for 90°? I've been looking through the web, and there's pretty much nothing for MKS Pi.

**Reply by nesickc** [reply]
Okay. I actually managed to find what I needed.

Basically, the gist is, we decompile the `rk3328-roc-cc.dtb` into `rk3328-roc-cc.dtc` with `dtc -I dtb -O dts -o rk3328-roc-cc.dts /boot/dtb/rockchip/rk3328-roc-cc.dtb` (I recommend making a new folder somewhere else for this, and backing up the `dtb` file).
Then we look for `rotate` in the `dts` file with ^W (if we're using nano).
When we find it, replace the number in the angled brackets with the angle we need (in hex). So if default is `rotate = < 0x10e >;`, for 90 degrees we go either `0x0`, or `0xb4`. For 180 degrees the example has rotated the image with `0x5a`.

Then we need to change the touch coordinate system. We do it by adding the transform matrix option in the `40-libinput.conf` file by running `sudo nano /usr/share/X11/xorg.conf.d/40-libinput.conf`.
Refer to https://klipperscreen.readthedocs.io/en/latest/Troubleshooting/Touch_issues/#save-touch-calibration for details.
There we do something like

```
Section "InputClass"
        Identifier "libinput touchscreen catchall"
        MatchIsTouchscreen "on"
        MatchDevicePath "/dev/input/event*"
        Driver "libinput"
        Option "TransformationMatrix" "0 -1 1 1 0 0 0 0 1"
EndSection
```

Then we do a reboot, and everything should work as expected.

For my usecase, I needed to rotate the screen clockwise 90 degrees, so I entered `rotate = < 0xb4 >;` and used the `0 -1 1 1 0 0 0 0 1` matrix. That's pretty much it

**Reply by XxyZwW** [reply]
I've already modified the decompiled .dts file, but do I need to recompile it into dtb? And how would I do that?
If I compiled it in a different folder as recommended, would I have to replace the .dtb file once I've recompiled it?

Thank you in advance for your help.

> Bien. De hecho, logré encontrar lo que necesitaba.
> 
> Básicamente, la esencia es que descompilamos el `rk3328-roc-cc.dtb` en `rk3328-roc-cc.dtc` con `dtc -I dtb -O dts -o rk3328-roc-cc.dts /boot/dtb/rockchip/rk3328-roc-cc.dtb` (Recomiendo hacer una nueva carpeta en otro lugar para esto y hacer una copia de seguridad del `dtb` archivo). Entonces buscamos `rotate` en el `dts` archivo con ^W (si estamos usando nano). Cuando lo encontremos, reemplace el número en los soportes en ángulo con el ángulo que necesitamos (en hexágono). Entonces, si el valor predeterminado es `rotate = < 0x10e >;`90 Grados vamos tampoco `0x0`, o `0xb4`. Para 180 grados el ejemplo ha girado la imagen con `0x5a`.
> 
> Entonces necesitamos cambiar el sistema de coordenadas táctiles. Lo hacemos agregando la opción de matriz de transformación en el `40-libinput.conf` archivo ejecutando `sudo nano /usr/share/X11/xorg.conf.d/40-libinput.conf`. Consulte https://klipperscreen.readthedocs.io/en/latest/Troubleshooting/Touch_issues/#save-touch-calibration para más detalles. Allí hacemos algo como
> 
> ```
> Section "InputClass"
>         Identifier "libinput touchscreen catchall"
>         MatchIsTouchscreen "on"
>         MatchDevicePath "/dev/input/event*"
>         Driver "libinput"
>         Option "TransformationMatrix" "0 -1 1 1 0 0 0 0 1"
> EndSection
> ```
> 
> Luego hacemos un reinicio, y todo debería funcionar como se esperaba.
> 
> Para mi caso de uso, necesitaba girar la pantalla en sentido horario 90 grados, así que entré `rotate = < 0xb4 >;` y usó el `0 -1 1 1 0 0 0 0 1` matriz. Eso es más o menos

**Reply by nesickc** [reply]
> I've already modified the decompiled .dts file, but do I need to recompile it into dtb? And how would I do that? If I compiled it in a different folder as recommended, would I have to replace the .dtb file once I've recompiled it?
> 
> Thank you in advance for your help.
> 
> > Bien. De hecho, logré encontrar lo que necesitaba.
> > Básicamente, la esencia es que descompilamos el `rk3328-roc-cc.dtb` en `rk3328-roc-cc.dtc` con `dtc -I dtb -O dts -o rk3328-roc-cc.dts /boot/dtb/rockchip/rk3328-roc-cc.dtb` (Recomiendo hacer una nueva carpeta en otro lugar para esto y hacer una copia de seguridad del `dtb` archivo). Entonces buscamos `rotate` en el `dts` archivo con ^W (si estamos usando nano). Cuando lo encontremos, reemplace el número en los soportes en ángulo con el ángulo que necesitamos (en hexágono). Entonces, si el valor predeterminado es `rotate = < 0x10e >;`90 Grados vamos tampoco `0x0`, o `0xb4`. Para 180 grados el ejemplo ha girado la imagen con `0x5a`.
> > Entonces necesitamos cambiar el sistema de coordenadas táctiles. Lo hacemos agregando la opción de matriz de transformación en el `40-libinput.conf` archivo ejecutando `sudo nano /usr/share/X11/xorg.conf.d/40-libinput.conf`. Consulte https://klipperscreen.readthedocs.io/en/latest/Troubleshooting/Touch_issues/#save-touch-calibration para más detalles. Allí hacemos algo como
> > ```
> > Section "InputClass"
> >         Identifier "libinput touchscreen catchall"
> >         MatchIsTouchscreen "on"
> >         MatchDevicePath "/dev/input/event*"
> >         Driver "libinput"
> >         Option "TransformationMatrix" "0 -1 1 1 0 0 0 0 1"
> > EndSection
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Luego hacemos un reinicio, y todo debería funcionar como se esperaba.
> > Para mi caso de uso, necesitaba girar la pantalla en sentido horario 90 grados, así que entré `rotate = < 0xb4 >;` y usó el `0 -1 1 1 0 0 0 0 1` matriz. Eso es más o menos

yes, you'd need to recompile it back. Sorry not mentioned it there. Once you recompile it you should replace the dtb file where you originally took it.
As for the command to compile, I don't remember exactly, but should be something like following:
`dtc -I dts -O dtb -o rk3328-roc-cc.dtb rk3328-roc-cc.dts`. Cannot check for sure because I sold that printer (it works well with no issues related to the screen, as far as the new owner is concerned)

**Comment by alfrix** [comment]
the manufacturer should provide the preferred way to rotate the screen, if that is not the case then there are good examples on how to rotate in the docs https://klipperscreen.readthedocs.io/en/latest/Hardware/#screen-rotation
